### ISSUE TITLE
Release the chart for v0.3 in the Rancher Apps

### ISSUE THREAD

**Issue by yasker** [rancher/longhorn#200]
1. It's incompatible to upgrade from v0.1 to v0.3.1 as a Rancher App.
2. Change the default namespace from `longhorn` to `longhorn-system` in v0.3. And make it impossible to deploy multiple Longhorn apps.
3. If possible, automatically determine the Flexvolume directory based on the distro of the cluster that Longhorn is going to be installed in.

**Comment by LLParse** [comment]
@yasker The only way to instruct Rancher on what default namespace to use is by the name of the chart. Even then, the next chart deployment will append a random suffix to the namespace. User can always modify the namespace. Renaming the chart to `longhorn-system` has an undesirable consequence: https://i.imgur.com/iL7b7P9.png

Alternatively, we can hard-code the correct namespace in the templates instead of letting Rancher decide. This has an undesirable consequence because it appears to be impossible to restrict charts from being deployed multiple times. There will be resource collisions if a user tries to deploy a second copy on the same cluster.

The last (and I think, best) alternative is to parameterize namespace in Longhorn to allow command-line configuration. Make all CRDs namespace scoped as well.

I spoke with the team and, based on what I've gathered, #2 and #3 aren't possible now but we can add the features in a future release. We'll also need a feature to specify that the chart can only run on Rancher 2.0.8 or later, (or whatever version the features are introduced in).

**Comment by yasker** [comment]
So IIUC:
1. Set the namespace statically to `longhorn-system` should prevent the user from automatically upgrade the system? Or the user would end up with two namespaces?
2. At least the statically namespace would stop the user from deploying multiple instances of the cluster.
3. OK, let's do the distro detection later. Probably it's nice if we can add this to our environment check script. Do you think that would be possible?

**Comment by LLParse** [comment]
Okay, so this took more testing than I expected.

1. There is no way to prevent the upgrade option from appearing besides renaming the new chart or deleting the old chart version. The upgrade 0.1.0->0.3.0 will fail and rollback doesn't seem to have an effect. The system remains usable, but stuck in `Installing` state indefinitely.
2. With the static namespace in place, the App name doesn't get automatically changed like it would otherwise. User gets a name conflict when they try to deploy a second instance. If user is really persistent, renaming the chart allows deployment. When deployed, the behavior becomes very strange. All of the resources show up under the new App and disappear from the old. At this point, two Apps are connected to the same resources. Deleting either one (you'll be inclined to delete the empty App) will tear down Longhorn.
3. I'm fairly certain we can script it

A more subtle issue appears when tearing down Longhorn. Unless every instance of every Longhorn CRD is deleted beforehand, CRD deletion gets stuck. Rancher also seems to explicitly terminate Pods (called `Workloads`) in addition to deleting the higher level resources. Consequently, our `driver-deployer` Deployment gets terminated before it can finish deleting CSI/flexvolume driver, resulting in orphaned Pods/Deployment. Re-deployment of any version Longhorn at this point yields non-deterministic behavior.

The TL;DR here is that we should also build a `helm` chart and test deployment/removal/upgrade with just `helm`. I think we've achieved all that is possible with the UI.

**Comment by LLParse** [comment]
If user does upgrade their Longhorn v0.1 system from Rancher UI and broke their Longhorn App as a result, they can follow the upgrade instructions from here: https://github.com/rancher/longhorn/blob/master/docs/upgrade.md#uninstall-the-old-version-of-longhorn

User needs to treat the release as v0.1 (not 0.3.0) even though it may appear as if the upgrade succeeded.

**Comment by yasker** [comment]
Done.
